demos/common/processing/async_pipeline.cpp

```diff
@@ -0,0 +1,217 @@
+    }
+
+    // --------------------------- 2. Read IR Generated by ModelOptimizer (.xml and .bin files) ------------
+    slog::info << "Loading network files" << slog::endl;
+    /** Read network model **/
+    InferenceEngine::CNNNetwork cnnNetwork = ie->ReadNetwork(model->getModelFileName());
+    /** Set batch size to 1 **/
+    slog::info << "Batch size is forced to 1." << slog::endl;
+    cnnNetwork.setBatchSize(1);
+
+    // -------------------------- Reading all outputs names and customizing I/O blobs (in inherited classes)
+    model->prepareInputsOutputs(cnnNetwork);
+
+    // --------------------------- 4. Loading model to the device ------------------------------------------
+    slog::info << "Loading model to the device" << slog::endl;
+    execNetwork = ie->LoadNetwork(cnnNetwork, cnnConfig.devices, cnnConfig.execNetworkConfig);
+
+    // --------------------------- 5. Create infer requests ------------------------------------------------
+    requestsPool.reset(new RequestsPool(execNetwork, cnnConfig.maxAsyncRequests));
+
+    // --------------------------- 6. Call onLoadCompleted to complete initialization of model -------------
+    model->onLoadCompleted(&execNetwork, requestsPool.get());
+}
+
+PipelineBase::~PipelineBase(){
+    waitForTotalCompletion();
+}
+
+
+void PipelineBase::waitForData(bool shouldKeepOrder){
+    std::unique_lock<std::mutex> lock(mtx);
+
+    condVar.wait(lock, [&] {return callbackException != nullptr ||
+        requestsPool->isIdleRequestAvailable() ||
+        (shouldKeepOrder ?
+            completedInferenceResults.find(outputFrameId) != completedInferenceResults.end() :
+            !completedInferenceResults.empty());
+    });
+
+    if (callbackException)
+        std::rethrow_exception(callbackException);
+}
+
+int64_t PipelineBase::submitRequest(InferenceEngine::InferRequest::Ptr request, std::shared_ptr<MetaData> metaData){
+    perfInfo.numRequestsInUse = (uint32_t)requestsPool->getInUseRequestsCount();
+
+    auto frameStartTime = std::chrono::steady_clock::now();
+
+    if (!perfInfo.startTime.time_since_epoch().count())
+    {
+        perfInfo.startTime = frameStartTime;
+    }
+
+    auto frameID = inputFrameId;
+
+    request->SetCompletionCallback([this,
+        frameStartTime,
+        frameID,
+        request,
+        metaData] {
+            {
+                std::lock_guard<std::mutex> lock(mtx);
+
+                try {
+                    InferenceResult result;
+
+                    result.frameId = frameID;
+                    result.metaData = std::move(metaData);
+                    for (std::string outName : model->getOutputsNames())
+                        result.outputsData.emplace(outName, std::make_shared<TBlob<float>>(*as<TBlob<float>>(request->GetBlob(outName))));
+                    result.startTime = frameStartTime;
+                    perfInfo.lastInferenceLatency = std::chrono::steady_clock::now() - frameStartTime;
```

Comment:
Why do you count time in the middle of callback? This means filling `result` is part of inference, but updating `PipelineBase`'s state isn't, which is strange.

Reply:
this is INFERENCE time latency. There's also general latency that is calculated right after getting processed data frame from the queue

Reply:
So what do you call inference? Why your definition includes filling `result` and doesn't include updating `PipelineBase`'s state?

Reply:
The inference latency here is time that inference engine takes to process request (time from submission till calling the callback). Thanks for noting - I've moved time calculation to the very start of callback to make it more precise. 
This metric is useful to check if we have performance issues in postprocessing/application code or in inference engine itself.

Reply:
will be changed after switching to PerformanceCounters

Reply:
done